File: mqtt/src/main/java/org/sheedon/mqtt/OkMqttClient.java
package org.sheedon.mqtt;

import android.content.Context;
import android.util.Log;

import org.eclipse.paho.android.service.MqttAndroidClient;
import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
import org.eclipse.paho.client.mqttv3.IMqttToken;
import org.eclipse.paho.client.mqttv3.MqttException;
import org.eclipse.paho.client.mqttv3.MqttMessage;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Queue;
import java.util.UUID;
import java.util.concurrent.LinkedBlockingDeque;

import static org.sheedon.mqtt.Util.checkNotNull;

/**
 * OkMqtt客户端，创建绑定反馈处理
 *
 * @Author: sheedon
 * @Email: sheedonsun@163.com
 * @Date: 2020/4/5 16:55
 */
public class OkMqttClient implements RealClient, MQTTFactory {
    private static final String TAG = "OK_MQTT_CLIENT";

    // 调度器
    private final Dispatcher dispatcher;
    // 超时毫秒
    private final long timeOutMilliSecond;
    // mqttAndroid客户端
    private final MqttAndroidClient mqttClient;

    // 反馈监听器
    private final MqttCallbackExtendedListener listener;

    // 连接选项配置类
    private final MqttConnectOptions connectOptions;
    // 断开连接缓存选项配置类
    private final DisconnectedBufferOptions disconnectedOptions;

    // 主题订阅数据
    private final Queue<SubscribeBody> subscribeBodies;

    // 基础主题
    private final String baseTopic;
    // 编码名称
    private final String charsetName;

    // 是否自动订阅主题
    private final boolean isAutoSubscribeToTopic;

    // 是否开始连接
    private boolean isStartConnect;

    public OkMqttClient() {
        this(new Builder());
    }

    private OkMqttClient(Builder builder) {
        this.dispatcher = builder.dispatcher;
        this.timeOutMilliSecond = builder.messageTimeout * 1000;
        this.charsetName = builder.charsetName;
        this.subscribeBodies = builder.subscribeBodies;
        this.baseTopic = builder.baseTopic;
        this.isAutoSubscribeToTopic = builder.isAutoSubscribeToTopic;
        this.dispatcher.setConverterFactories(Collections.unmodifiableList(builder.converterFactories));

        checkNotNull(builder.context, "context is null");
        checkNotNull(builder.serverURI, "server uri is null");

        String clientId = builder.clientId == null || builder.clientId.equals("") ? UUID.randomUUID().toString() : builder.clientId;
        mqttClient = new MqttAndroidClient(builder.context, builder.serverURI, clientId);
        this.listener = builder.listener;
        mqttClient.setCallback(mCreateCallback);


        this.connectOptions = builder.connectOptions;
        this.disconnectedOptions = builder.disconnectedOptions;

        connect();
    }

    public Builder newBuilder() {
        return new Builder();
    }

    /**
     * mqtt 创建连接
     */
    private void connect() {
        checkNotNull(mqttClient, "please create mqtt client");
        if (!mqttClient.isConnected() && !isStartConnect) {
            try {
                isStartConnect = true;
                mqttClient.connect(connectOptions, null, mCreateCallback);
            } catch (MqttException e) {
                e.printStackTrace();
            }
        }
    }

    private long lastTime = 0;

    public void reConnect() {
        long nowTime = System.currentTimeMillis();
        if (nowTime - lastTime < 5000) {
            isStartConnect = false;
            if (listener != null)
                listener.onFailure(null, new Throwable("Only reconnect once within 5 seconds"));
            return;
        }

        lastTime = nowTime;
        try {
            connect();
        } catch (Exception e) {
            if (listener != null)
                listener.onFailure(null, e);
        }
    }

    /**
     * 获取调度器
     */
    public Dispatcher dispatcher() {
        return dispatcher;
    }

    /**
     * 获取超时毫秒数
     */
    public long timeOutMilliSecond() {
        return timeOutMilliSecond;
    }

    /**
     * mqttAndroid客户端
     */
    public MqttAndroidClient mqttClient() {
        return mqttClient;
    }

    /**
     * 基础主题
     */
    String baseTopic() {
        return baseTopic;
    }

    /**
     * mqtt连接监听
     */
    private final MqttCallbackExtendedListener mCreateCallback = new MqttCallbackExtendedListener() {

        /**
         * 连接成功
         * */
        @Override
        public void onSuccess(IMqttToken asyncActionToken) {

            isStartConnect = false;
            mqttClient.setBufferOpts(disconnectedOptions);
            if (listener != null)
                listener.onSuccess(asyncActionToken);
        }

        /**
         * 连接失败
         * */
        @Override
        public void onFailure(IMqttToken asyncActionToken, Throwable exception) {
            isStartConnect = false;

            if (listener != null)
                listener.onFailure(asyncActionToken, exception);
        }

        @Override
        public void connectComplete(boolean reconnect, String serverURI) {

            if (isAutoSubscribeToTopic)
                subscribeToTopic(reconnect);


            if (listener != null)
                listener.connectComplete(reconnect, serverURI);
        }

        @Override
        public void connectionLost(Throwable cause) {

            if (listener != null)
                listener.connectionLost(cause);
        }

        @Override
        public void messageArrived(String topic, MqttMessage message) throws Exception {

            Log.v(TAG, "messageArrived: " + topic + " : " + new String(message.getPayload(), charsetName));
            // 数据转义
            String data = new String(message.getPayload(), charsetName);

            dispatcher.enqueueNetCallback(topic, message, charsetName);
            messageArrived(topic, data);
        }

        @Override
        public void deliveryComplete(IMqttDeliveryToken token) {

            if (listener != null)
                listener.deliveryComplete(token);
        }

        @Override
        public void messageArrived(String topic, String data) {
            if (listener != null)
                listener.messageArrived(topic, data);
        }
    };


    /**
     * 订阅主题
     *
     * @param reconnect 是否重连
     */
    private void subscribeToTopic(boolean reconnect) {
        if (mqttClient == null
                || subscribeBodies == null
                || subscribeBodies.size() == 0)
            return;

        // 配置为自动清除 或者 非清除+第一次连接
        if ((!connectOptions.isCleanSession() && !reconnect)
                || connectOptions.isCleanSession()) {
            String[] topicFilters = new String[subscribeBodies.size()];
            int[] qos = new int[subscribeBodies.size()];
            int index = 0;
            for (SubscribeBody body : subscribeBodies) {
                topicFilters[index] = body.getTopic();
                qos[index] = body.getQos();
                index++;
            }
            try {
                mqttClient.subscribe(topicFilters, qos, null, listener);
            } catch (MqttException e) {
                e.printStackTrace();
            }
        }

    }

    public Queue<SubscribeBody> getSubscribeBodies() {
        return new LinkedBlockingDeque<>(subscribeBodies);
    }

    public void updateSubscribeBodies(Queue<SubscribeBody> queue) {
        if (subscribeBodies != null) {
            subscribeBodies.clear();
            subscribeBodies.addAll(queue);

            subscribeToTopic(true);
        }
    }

    /**
     * 创建Call
     *
     * @param request 请求数据
     * @return Call
     */
    @Override
    public Call newCall(Request request) {
        return RealCall.newRealCall(this, request);
    }

    /**
     * 创建观察者
     *
     * @param request 请求数据
     * @return Observable
     */
    @Override
    public Observable newObservable(Request request) {
        return RealObservable.newRealObservable(this, request);
    }


    public static final class Builder {
        Dispatcher dispatcher;

        Context context;
        String serverURI;
        String clientId;

        String baseTopic;

        MqttConnectOptions connectOptions;
        DisconnectedBufferOptions disconnectedOptions;

        int messageTimeout;

        Queue<SubscribeBody> subscribeBodies;
        boolean isAutoSubscribeToTopic;

        MqttCallbackExtendedListener listener;

        String charsetName;

        final List<DataConverter.Factory> converterFactories = new ArrayList<>();


        public Builder() {
            dispatcher = new Dispatcher();
            messageTimeout = 5;
            isAutoSubscribeToTopic = true;
            connectOptions = MqttConnectOptions.getDefault();
            disconnectedOptions = DisconnectedBufferOptions.getDefault();
            charsetName = "GBK";
            baseTopic = "";
        }

        /**
         * 设置用于设置策略和执行异步请求的调度程序。不能为null。
         */
        public Builder dispatcher(Dispatcher dispatcher) {
            if (dispatcher == null) throw new IllegalArgumentException("dispatcher == null");
            this.dispatcher = dispatcher;
            return this;
        }

        /**
         * 设置mqtt客户端基本信息。Context/serverURI不能为null。
         */
        public Builder clientInfo(Context context, String serverURI, String clientId) {
            if (context == null) throw new IllegalArgumentException("context == null");
            if (serverURI == null || serverURI.equals(""))
                throw new IllegalArgumentException("serverURI == null");
            if (clientId == null) throw new IllegalArgumentException("clientId == null");
            this.context = context;
            this.serverURI = serverURI;
            this.clientId = clientId;
            return this;
        }

        /**
         * 设置mqtt连接选项内容配置
         */
        public Builder mqttOptions(org.eclipse.paho.client.mqttv3.MqttConnectOptions options) {
            return mqttOptions((MqttConnectOptions) options);
        }

        /**
         * 设置mqtt连接选项内容配置
         */
        public Builder mqttOptions(MqttConnectOptions options) {
            if (options == null) throw new IllegalArgumentException("options == null");
            MqttConnectOptions.setDefault(options);
            return this;
        }

        /**
         * 设置mqtt连接选项内容配置
         */
        public Builder disconnectedOptions(org.eclipse.paho.client.mqttv3.DisconnectedBufferOptions options) {
            return disconnectedOptions((DisconnectedBufferOptions) options);
        }

        /**
         * 设置mqtt连接选项内容配置
         */
        public Builder disconnectedOptions(DisconnectedBufferOptions options) {
            if (options == null) throw new IllegalArgumentException("options == null");
            DisconnectedBufferOptions.setDefault(options);
            return this;
        }

        /**
         * 设置用户名和密码
         */
        public Builder userNameAndPassword(String userName, String password) {
            connectOptions.setUserName(userName);
            connectOptions.setPassword(password.toCharArray());
            return this;
        }

        /**
         * 基础主题- 用于请求数据时没有数据时添加
         */
        public Builder baseTopic(String baseTopic) {
            this.baseTopic = baseTopic;
            return this;
        }

        /**
         * 设置will主题
         */
        public Builder will(String topic, String payload) {
            return will(topic, payload, 1, true);
        }

        /**
         * 设置will主题
         */
        public Builder will(String topic, String payload, int qos, boolean retained) {
            connectOptions.setWill(topic, payload.getBytes(), qos, retained);
            return this;
        }

        /**
         * 设置保持在线的判断时间间隔
         */
        public Builder keepAliveInterval(int keepAliveInterval) {
            if (keepAliveInterval < 0)
                return this;

            connectOptions.setKeepAliveInterval(keepAliveInterval);
            return this;
        }

        /**
         * 设置连接超时时间
         */
        public Builder connectionTimeout(int connectionTimeout) {
            if (connectionTimeout < 0)
                return this;

            connectOptions.setConnectionTimeout(connectionTimeout);
            return this;
        }

        /**
         * 设置信息请求超时时间
         */
        public Builder messageTimeout(int messageTimeout) {
            if (messageTimeout < 0)
                return this;

            this.messageTimeout = messageTimeout;
            return this;
        }

        /**
         * 设置最大并发值
         */
        public Builder maxInflight(int maxInflight) {
            if (maxInflight < 0)
                return this;

            connectOptions.setMaxInflight(maxInflight);
            return this;
        }

        /**
         * 设置客户端和服务器是否应该记住重新启动和重新连接之间的状态。
         */
        public Builder cleanSession(boolean cleanSession) {
            connectOptions.setCleanSession(cleanSession);
            return this;
        }

        /**
         * 是否自动订阅主题
         */
        public Builder isAutoSubscribeToTopic(boolean isAutoSubscribeToTopic) {
            this.isAutoSubscribeToTopic = isAutoSubscribeToTopic;
            return this;
        }

        /**
         * 设置如果连接断开，客户端是否将自动尝试重新连接到服务器
         */
        public Builder automaticReconnect(boolean automaticReconnect) {
            connectOptions.setAutomaticReconnect(automaticReconnect);
            return this;
        }

        /**
         * 设置订阅主题内容
         */
        public Builder subscribeBodies(Queue<SubscribeBody> subscribeBodies) {
            this.subscribeBodies = subscribeBodies;
            return this;
        }

        /**
         * 添加订阅主题内容
         */
        public Builder addSubscribeBodies(Queue<SubscribeBody> subscribeBodies) {
            if (this.subscribeBodies == null)
                this.subscribeBodies = new ArrayDeque<>();

            this.subscribeBodies.addAll(subscribeBodies);
            return this;
        }

        /**
         * 添加订阅主题内容
         */
        public Builder addSubscribeBodies(SubscribeBody subscribeBody) {
            if (this.subscribeBodies == null)
                this.subscribeBodies = new ArrayDeque<>();

            this.subscribeBodies.add(subscribeBody);
            return this;
        }

        /**
         * 设置反馈监听器
         */
        public Builder callback(MqttCallbackExtendedListener listener) {
            this.listener = listener;
            return this;
        }

        /**
         * 设置字符转义
         */
        public Builder charsetName(String charsetName) {
            this.charsetName = charsetName;
            return this;
        }


        /**
         * 消息处理工具
         */
        public Builder addConverterFactory(DataConverter.Factory factory) {
            converterFactories.add(checkNotNull(factory, "DataConverter.MqttFactory == null"));
            return this;
        }


        public OkMqttClient build() {

            if (converterFactories.size() == 0) {
                throw new IllegalStateException("converterFactories is null.");
            }

            return new OkMqttClient(this);
        }
    }
}

File: mqtt/src/main/java/org/sheedon/mqtt/OkMqttClient.kt
package org.sheedon.mqtt

import android.content.Context
import org.eclipse.paho.android.service.MqttAndroidClient
import org.eclipse.paho.client.mqttv3.*
import org.sheedon.mqtt.listener.*
import org.sheedon.rr.core.*
import org.sheedon.rr.dispatcher.Dispatcher
import org.sheedon.rr.timeout.TimeoutManager

/**
 * Mqtt 门面客户端类，包含「请求响应客户端」和「mqtt调度客户端」
 *
 * @Author: sheedon
 * @Email: sheedonsun@163.com
 * @Date: 2022/1/26 10:40 上午
 */
class OkMqttClient internal constructor(
    builder: Builder
) {

    private val mqttRRBinderClient: MqttRRBinderClient = builder.mqttRRBinderClient!!
    val mqttClient: MqttWrapperClient = builder.mqttClient!!

    init {
        val requestAdapter = mqttRRBinderClient.dispatchManager.requestAdapter()
        requestAdapter!!.bindSender(mqttClient)

        mqttClient.bindDispatchAdapter(mqttRRBinderClient.switchMediator)

    }

    /**
     * 创建请求响应的Call
     *
     * @param request 请求对象
     * @return Call 用于执行入队/提交请求的动作
     */
    fun newCall(request: IRequest<String, RequestBody>): Call {
        return mqttRRBinderClient.newCall(request)
    }

    /**
     * 创建信息的观察者 Observable
     *
     * @param request 请求对象
     * @return Observable 订阅某个主题，监听该主题的消息
     */
    fun newObservable(request: IRequest<String, RequestBody>): Observable {
        return mqttRRBinderClient.newObservable(request)
    }

    /**
     * 重连操作
     */
    @JvmOverloads
    fun reConnect(
        listener: IResultActionListener? = null
    ) {
        this.mqttClient.reConnect(listener)
    }

    @JvmOverloads
    fun disConnect(listener: IResultActionListener? = null) {
        this.mqttClient.disConnect(listener)
    }


    /**
     * 订阅mqtt主题
     * @param body mqtt消息体
     * @param attachRecord 是否附加到缓存记录中，若false，则代表单次订阅，清空行为后，不恢复
     * @param listener 操作监听器
     */
    @JvmOverloads
    fun subscribe(
        body: SubscribeBody,
        attachRecord: Boolean = false,
        listener: IResultActionListener? = null
    ) {
        this.mqttClient.subscribe(body, attachRecord, listener)
    }

    /**
     * 订阅mqtt主题
     * @param bodies mqtt消息体集合
     * @param attachRecord 是否附加到缓存记录中，若false，则代表单次订阅，清空行为后，不恢复
     * @param listener 操作监听器
     */
    @JvmOverloads
    fun subscribe(
        bodies: List<SubscribeBody>,
        attachRecord: Boolean = false,
        listener: IResultActionListener? = null
    ) {
        this.mqttClient.subscribe(bodies, attachRecord, listener)
    }

    /**
     * 取消订阅mqtt主题
     * @param body mqtt消息体
     * @param attachRecord 是否附加到缓存记录中，若false，则代表单次订阅，清空行为后，不恢复
     * @param listener 操作监听器
     */
    @JvmOverloads
    fun unsubscribe(
        body: SubscribeBody,
        attachRecord: Boolean = false,
        listener: IResultActionListener? = null
    ) {
        this.mqttClient.unsubscribe(body, attachRecord, listener)
    }

    /**
     * 取消订阅mqtt主题
     * @param bodies mqtt消息体集合
     * @param attachRecord 是否附加到缓存记录中，若false，则代表单次订阅，清空行为后，不恢复
     * @param listener 操作监听器
     */
    @JvmOverloads
    fun unsubscribe(
        bodies: List<SubscribeBody>,
        attachRecord: Boolean = false,
        listener: IResultActionListener? = null
    ) {
        this.mqttClient.unsubscribe(bodies, attachRecord, listener)
    }

    /**
     * 重新订阅mqtt主题
     *
     * @param bodies mqtt消息体集合
     * @param listener 操作监听器
     */
    @JvmOverloads
    fun reSubscribe(
        bodies: List<SubscribeBody>,
        listener: IResultActionListener? = null
    ) {
        this.mqttClient.reSubscribe(bodies, listener)
    }

    /**
     * 发送mqtt消息
     *
     * @param topic 主题
     * @param message mqtt消息内容
     */
    fun publish(topic: String, message: MqttMessage): IMqttDeliveryToken {
        return mqttClient.publish(topic, message)
    }


    class Builder {
        internal var mqttRRBinderClient: MqttRRBinderClient? = null
        internal var mqttClient: MqttWrapperClient? = null

        internal val mqttRRBinderBuilder: MqttRRBinderClient.Builder = MqttRRBinderClient.Builder()
        internal val mqttBuilder: MqttWrapperClient.Builder = MqttWrapperClient.Builder()

        /**
         * 设置 mqtt 请求响应绑定客户端
         * @param mqttRRBinderClient 请求响应绑定客户端
         */
        fun mqttRRBinderClient(mqttRRBinderClient: MqttRRBinderClient) = apply {
            this.mqttRRBinderClient = mqttRRBinderClient
        }

        /**
         * 设置 mqtt调度客户端
         * @param mqttClient mqtt调度客户端
         */
        fun mqttClient(mqttClient: MqttWrapperClient) = apply {
            this.mqttClient = mqttClient
        }

        /**
         * 设置基础主题，后续添加的topic 则在此基础上拼接
         *
         * @param baseTopic 基础主题
         * @return Builder 构建者
         */
        fun baseTopic(baseTopic: String) = apply {
            this.mqttRRBinderBuilder.baseTopic(baseTopic)
        }

        /**
         * 设置字符集编码类型，在接收数据时转化为指定格式的字符串
         *
         * @param charsetName 字符集编码类型
         * @return Builder 构建者
         */
        fun charsetName(charsetName: String) = apply {
            this.mqttRRBinderBuilder.charsetName(charsetName)
        }

        /**
         * 设置用于设置策略和执行异步请求的调度程序。不能为null。
         *
         * @param dispatcher 请求响应执行者
         */
        fun dispatcher(dispatcher: Dispatcher<String, String, RequestBody, ResponseBody>) = apply {
            this.mqttRRBinderBuilder.dispatcher(dispatcher)
        }

        /**
         * 设置信息请求超时时间（单位秒）
         *
         * @param timeout 超时时间
         * @return Builder<BackTopic></BackTopic>, ID> 构建者
         */
        fun messageTimeout(timeout: Int) = apply {
            if (timeout < 0) return this
            this.mqttRRBinderBuilder.messageTimeout(timeout)
        }

        /**
         * 设置行为服务环境集合
         *
         * @param behaviorServices 执行服务环境集合
         * @return Builder<BackTopic></BackTopic>, ID>
         */
        fun behaviorServices(behaviorServices: MutableList<EventBehavior>) = apply {
            this.mqttRRBinderBuilder.behaviorServices(behaviorServices)
        }

        /**
         * 设置行为线程池，后加的靠前
         *
         * @param behaviorService 执行服务环境
         * @return Builder<BackTopic></BackTopic>, ID>
         */
        fun behaviorService(behaviorService: EventBehavior) = apply {
            this.mqttRRBinderBuilder.behaviorService(behaviorService)
        }

        /**
         * 设置事件管理集合
         *
         * @param eventManagerPool 事件管理集合
         * @return Builder<BackTopic></BackTopic>, ID>
         */
        fun eventManagerPool(eventManagerPool: MutableList<EventManager<String, String, RequestBody, ResponseBody>>) =
            apply {
                this.mqttRRBinderBuilder.eventManagerPool(eventManagerPool)
            }

        /**
         * 设置事件管理
         *
         * @param eventManager 事件管理
         * @return Builder<BackTopic></BackTopic>, ID>
         */
        fun eventManager(eventManager: EventManager<String, String, RequestBody, ResponseBody>) =
            apply {
                this.mqttRRBinderBuilder.eventManager(eventManager)
            }

        /**
         * 设置超时处理者
         *
         * @param timeoutManager 事件管理
         * @return Builder<BackTopic></BackTopic>, ID>
         */
        fun timeoutManager(timeoutManager: TimeoutManager<String>) = apply {
            this.mqttRRBinderBuilder.timeoutManager(timeoutManager)
        }

        /**
         * 设置调度调度适配者
         *
         * @param dispatchAdapter 调度适配者
         * @return Builder<BackTopic></BackTopic>, ID>
         */
        fun dispatchAdapter(dispatchAdapter: DispatchAdapter<RequestBody, ResponseBody>) = apply {
            this.mqttRRBinderBuilder.dispatchAdapter(dispatchAdapter)
        }

        /**
         * 设置请求适配者
         *
         * @param requestAdapter 请求适配者
         * @return Builder<BackTopic></BackTopic>, ID>
         */
        fun requestAdapter(requestAdapter: RequestAdapter<RequestBody>) = apply {
            this.mqttRRBinderBuilder.requestAdapter(requestAdapter)
        }

        /**
         * 设置反馈主题转换器集合
         *
         * @param backTopicConverters 反馈主题转换器集合
         * @return Builder<BackTopic></BackTopic>, ID>
         */
        fun backTopicConverters(backTopicConverters: MutableList<DataConverter<ResponseBody, String>>) =
            apply {
                this.mqttRRBinderBuilder.backTopicConverters(backTopicConverters)
            }

        /**
         * 设置反馈主题转换器
         *
         * @param backTopicConverter 反馈主题转换器
         * @return Builder<BackTopic></BackTopic>, ID>
         */
        fun addBackTopicConverter(backTopicConverter: DataConverter<ResponseBody, String>) =
            apply {
                this.mqttRRBinderBuilder.addBackTopicConverter(backTopicConverter)
            }

        /**
         * 设置响应调度适配者
         *
         * @param responseAdapter 响应调度适配者
         * @return Builder<BackTopic></BackTopic>, ID>
         */
        fun responseAdapter(responseAdapter: ResponseAdapter<String, ResponseBody>) = apply {
            this.mqttRRBinderBuilder.responseAdapter(responseAdapter)
        }

        /**
         * Enables an android application to communicate with an MQTT server using non-blocking methods.
         *
         * @param androidClient MqttAndroidClient
         * @return Builder
         */
        fun androidClient(androidClient: MqttAndroidClient) = apply {
            this.mqttBuilder.androidClient(androidClient)
        }

        /**
         * Constructor- create an MqttAndroidClient that can be used to communicate
         * with an MQTT server on android
         *
         * @param context     used to pass context to the callback.
         * @param serverURI   specifies the protocol, host name and port to be used to
         * connect to an MQTT server
         * @param clientId    specifies the name by which this connection should be
         * identified to the server
         * @param persistence the persistence class to use to store in-flight message. If
         * null then the default persistence mechanism is used
         * @param ackType     how the application wishes to acknowledge a message has been
         * processed.
         * @return Builder
         */
        @JvmOverloads
        fun clientInfo(
            context: Context,
            serverURI: String,
            clientId: String?,
            persistence: MqttClientPersistence? = null,
            ackType: MqttAndroidClient.Ack? = MqttAndroidClient.Ack.AUTO_ACK
        ) = apply {
            this.mqttBuilder.clientInfo(context, serverURI, clientId, persistence, ackType)
        }

        /**
         * set of connection parameters that override the defaults
         *
         * @param options  connection parameters
         * @param <Option> MqttConnectOptions
         * @return Builder
        </Option> */
        fun <Option : MqttConnectOptions> connectOptions(options: Option) = apply {
            this.mqttBuilder.connectOptions(options)
        }

        /**
         * Sets the DisconnectedBufferOptions for this client
         *
         * @param bufferOpts the DisconnectedBufferOptions
         * @param <Option>   DisconnectedBufferOptions
         * @return Builder
         */
        fun <Option : DisconnectedBufferOptions> disconnectedBufferOptions(bufferOpts: Option) =
            apply {
                this.mqttBuilder.disconnectedBufferOptions(bufferOpts)
            }

        /**
         * Sets the MessageListener for this client
         *
         * @param callback the MessageListener
         * @return Builder
         */
        fun callback(callback: IMessageListener) = apply {
            this.mqttBuilder.callback(callback)
        }

        /**
         * Sets the MessageListener for this client
         *
         * @param connectListener the connect action listener
         * @return Builder
         */
        @JvmOverloads
        fun connectListener(
            connectListener: IActionListener? = null
        ) = apply {
            this.mqttBuilder.connectListener(connectListener)
        }

        /**
         * Sets the default subscribe info for this client and bind listener
         *
         * @param subscribeBodies Subscribe topic and qos
         * @param subscribeListener The subscribe action listener
         * @param autoSubscribe Whether to subscribe automatically after reconnect
         * @return Builder
         */
        @JvmOverloads
        fun subscribeBodies(
            subscribeListener: IActionListener? = null,
            autoSubscribe: Boolean = false,
            vararg subscribeBodies: SubscribeBody
        ) = apply {
            this.mqttBuilder.subscribeBodies(subscribeListener, autoSubscribe, *subscribeBodies)
        }

        fun build(): OkMqttClient {
            if (mqttRRBinderClient == null) {
                mqttRRBinderClient = mqttRRBinderBuilder.build()
            }
            if (mqttClient == null) {
                mqttClient = mqttBuilder.build()
            }

            return OkMqttClient(this)
        }
    }

}